examples/allocation-endpoint/README.md

```diff
@@ -0,0 +1,104 @@
+
+<pre>
+gcloud container clusters create [NAME] --cluster-version=1.21 \
+  --project=[GKE-PROJECT-ID] \
+  --tags=game-server \
+  --scopes=gke-default \
+  --num-nodes=4 \
+  --machine-type=e2-standard-4 \
+  <b>--workload-pool=[GKE-PROJECT-ID].svc.id.goog \ </b>
+  --region=[REGION]
+</pre>
+
+Install Agones on your cluster. You need to disable mTLS because the `agones-allocator` container will be the backend for ESP container.
+
+```
+helm upgrade my-release --install --namespace agones-system --create-namespace agones/agones \
+  --set agones.allocator.disableMTLS=true \
+  --set agones.allocator.disableTLS=true \
+  --set agones.allocator.service.http.enabled=false
+```
+
+After installing Agones, deploy [ESP](https://cloud.google.com/endpoints/docs/grpc/specify-esp-v2-startup-options) which is an envoy based proxy in front of agones `agones-alloator` container. Run the following to patch the service deployement, change the service port to ESP and add annotation to `agones-allocator` service account to impersonate GCP service account. 
+
+Replace [GKE-PROJECT-ID] in `patch-agones-allocator.yaml` with your project ID before runing the scripts.
+
+```
+kubectl patch deployment agones-allocator -n agones-system --patch-file patch-agones-allocator.yaml
+kubectl patch svc agones-allocator -n agones-system --type merge -p '{"spec":{"ports": [{"port": 443,"name":"https","targetPort":9443}]}}'
+kubectl annotate sa -n agones-system agones-allocator iam.gke.io/gcp-service-account=ae-esp-sa@[PROJECT-ID].iam.gserviceaccount.com
+```
+
+## Terraform 
+The terraform modules creates resources in GCP:
+
+```
+terraform apply -var "project_id=[PROJECT-ID]" -var "ae_proxy_image=[GCR-IMAGE]" -var "authorized_members=[\"serviceAccount:[SERVICE-ACCOUNT-EMAIL]\"]" -var "clusters_info=[CLUSTERS-INFO]" -var "workload-pool=[GKE-PROJECT-ID].svc.id.goog"
+```
+
+`[CLUSTERS-INFO]` is in the form of `[{\"name\":\"cluster1\",\"endpoint\":\"34.83.14.82\",\"namespace\":\"default\",\"allocation_weight\":100},{...}]` deserializing to []ClusterInfo, defined in the `server/clusterselector.go`.
+
+  The `name` is a unique randomly selected name for the cluster. 
+  The `endpoint` is the `agones-allocator` external IP. 
+  The `namespace` is the game server namespace.  
+  The `allocation_weight` is a value between 0 and 100, which sets the relative allocation rate a cluster receives compared to other clusters. By setting weight to zero, a cluster stops receiving allocation requests.
+
+`[GCR-IMAGE]` has a default set to `gcr.io/agones-images/allocation-endpoint-proxy`.
+
+`[SERVICE-ACCOUNT-EMAIL]` is the service account to be granted access the Allocation Endpoint.
+
+## Server
+
+The Allocation Endpoint proxy code is in `./server` folder. You can make changes and run the following to build and push the image to your own GCR repository:
+
+```
+docker build --tag gcr.io/[PROJECT-ID]/allocation-endpoint-proxy .
+docker push gcr.io/[PROJECT-ID]/allocation-endpoint-proxy
```

Comment:
Should this have a version tag?

Reply:
In the other example directories, we specify a version. It makes it clear that something has changed in the source (to warrant the version being bumped). 

I'm not particularly familiar with how cloud run works, but can you specify your image as `gcr.io/[PROJECT-ID]/allocation-endpoint-proxy:latest` when running the proxy?

Reply:
Yes, you can add image tag. if the tag is not specified it defaults to the latest. To clarify I added the tag.